File: web/lib/customItems.ts
// Custom meal entries store items newline-separated in `customName`, each
// `name[<TAB>amount[<TAB>unit]]` (tab can't be typed in a text input).

export interface CustomItem {
  name: string
  amount: string
  unit?: string
  // UI-only: saved to the name->category catalog, never encoded.
  categoryId?: string
}

const SEP = '\t'

export function parseCustomItems(customName: string): CustomItem[] {
  return customName
    .split('\n')
    .filter(Boolean)
    .map((line) => {
      const parts = line.split(SEP)
      if (parts.length === 1) return { name: parts[0], amount: '' }
      if (parts.length === 2) return { name: parts[0], amount: parts[1] }
      return { name: parts[0], amount: parts[1], unit: parts[2] }
    })
}

export function encodeCustomItems(items: CustomItem[]): string {
  return items
    .map((it) => ({ name: it.name.trim(), amount: it.amount.trim(), unit: (it.unit ?? '').trim() }))
    .filter((it) => it.name)
    .map((it) => {
      if (!it.amount) return it.name
      if (!it.unit) return `${it.name}${SEP}${it.amount}`
      return `${it.name}${SEP}${it.amount}${SEP}${it.unit}`
    })
    .join('\n')
}

export function formatCustomItemLabel(item: CustomItem): string {
  if (!item.amount) return item.name
  const unit = item.unit?.trim()
  return unit ? `${item.amount} ${unit} ${item.name}` : `${item.amount} ${item.name}`
}

export function formatCustomNameLabel(customName: string): string {
  return parseCustomItems(customName).map(formatCustomItemLabel).join('\n')
}

File: api/apps/mealplans/custom_items.go
package mealplans

import "strings"

// customItemSep separates a custom item's name from its amount ("apples\t2");
// a tab can't be typed into the single-line input.
const customItemSep = "\t"

// displayCustomName renders "name\tamount" lines as "amount name".
func displayCustomName(customName string) string {
	lines := strings.Split(customName, "\n")
	for i, line := range lines {
		name, amount, found := strings.Cut(line, customItemSep)
		if found && strings.TrimSpace(amount) != "" {
			lines[i] = strings.TrimSpace(amount) + " " + name
		} else {
			lines[i] = name
		}
	}
	return strings.Join(lines, "\n")
}
